refactor: types

hongaar opened the pull request:
Started with typing everything with TS strict mode on.

Question: what's the purpose of having additional `IVorpal` / `ICommand` interfaces, as they seem to largely shadow the implementing classes?

I think it would save a lot of time maintaining the lib if we could use the classes as reference types and get rid of the interfaces.

- https://github.com/dthree/vorpal/wiki
- https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/vorpal/index.d.ts
- https://github.com/dthree/vorpal/pull/272/files#diff-3a699a0719518f967e9b2df3f5549e02

## Changes

- Removing all function return types (let TS infer them instead)
- Removing all class interfaces (classes _are_ interfaces to TS)
- Typing all explicit/implicit any's
- Minor syntax upgrades (e.g. `_.isFunction(x)` --> `typeof x === 'function'` and `extend(a, b)` --> `{...a, ...b}` etc.) 
- Adding type guards where needed (e.g. `if (matches === null) { throw new Error(...) }` etc.)
- Typing EventEmitter events

## Todo

### Misc

- [x] Fix tests

### Type

- [x] autocomplete-utils
- [x] autocomplete
- [x] command-instance
- [x] command
- [x] history
- [x] intercept
- [x] local-storage
- [x] logger
- [x] option
- [x] session
- [x] ui
- [x] util
- [x] vorpal-commons
- [x] vorpal

```
[Oct 28] Found 361 errors. Watching for file changes.
[Nov 25] Found 233 errors. Watching for file changes.
[Nov 26] Found 211 errors. Watching for file changes.
[Nov 30] Found 158 errors. Watching for file changes.
[Dec 1] Found 73 errors. Watching for file changes.
[Dec 10] Found 0 errors. Watching for file changes.
```

hongaar committed:
chore: start work to comply with strict mode [wip]

hongaar committed:
chore(types): wip

AdrieanKhisbe commented on .eslintrc.js:
```diff
@@ -0,0 +1,130 @@
+  plugins: ['@typescript-eslint'],
+  extends: ['prettier', 'prettier/@typescript-eslint'],
+  rules: {
+    '@typescript-eslint/adjacent-overload-signatures': 'error',
```
maybe we could go for using `plugin:@typescript-eslint/recommended` configuration, rather than configuring ourself each rules?
(and still overriding recommended config when it's not relally suited to what we want)

What do you think @vorpaljs-reforged/vorpal-developers ?

Ore4444 replied:
i agree our config should only include our explicit overrrides

AdrieanKhisbe commented on .travis.yml:
```diff
@@ -1,19 +1,19 @@
+node_js:
+  - "8"
+  - "10"
+  - "11"
```
```suggestion
  - "12"
```

Ore4444 replied:
merged this change already

AdrieanKhisbe commented on .travis.yml:
```diff
@@ -1,19 +1,19 @@
+
+script:
+  - npm run compile
+  - npm run test
```
I would prefer the `test` to contain the `lint` part first

hongaar replied:
Never use lint step like that, but I guess it would work 😄

AdrieanKhisbe commented on .travis.yml:
```diff
@@ -1,19 +1,19 @@
+  - npm run publish-coverage
+
+
+cache:
```
we could simplify all this block by `cache: npm`

AdrieanKhisbe commented on examples/calculator/calculator.js:
```diff
@@ -1,4 +1,5 @@
-var vorpal = require('./../../')();
+var Vorpal = require('../../dist/vorpal').default;
```
why the default?

AdrieanKhisbe commented on examples/calculator/calculator.js:
```diff
@@ -17,21 +18,19 @@ vorpal.command('add [numbers...]', 'Adds numbers together')
-        cb(undefined, 'double 2');
+  .autocomplete({
+      data: function (text, cb) {
+          var number = String(text).trim();
```
(make autocomplete Promise support comes quick :))

AdrieanKhisbe commented on package.json:
```diff
@@ -26,18 +26,18 @@
     "watch": "tsc -watch -p ./",
-    "lint:base": "tslint -p tsconfig.json",
+    "lint:base": "eslint --ext .ts src/**",
     "lint:check": "npm run lint:base",
```
would simplify we a simple `lint` and a `lint:fix`
```json
  "lint": "eslint --ext .ts src/**",
  "lint:fix": "npm run lint -- --fix",
```

AdrieanKhisbe commented on package.json:
```diff
@@ -26,18 +26,18 @@
+    "lint:base": "eslint --ext .ts src/**",
     "lint:check": "npm run lint:base",
     "lint:fix": "npm run lint:base -- --fix",
     "prettier:base": "prettier \"{.,{src,test}/**}/*.@(ts|js)\"",
```
not sure we need anymore with eslint integration

AdrieanKhisbe reviewed (dismissed):
First batch of suggestions.

I stoped when realized that for some reason this PR contains work from other PR.

hongaar commented:
> I stoped when realized that for some reason this PR contains work from other PR.

Rewrote history (was doing 6 things at once). It's better now 👍

Ore4444 commented on tsconfig.json:
```diff
@@ -1,42 +1,16 @@
-    // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
-    // "typeRoots": [],                       /* List of folders to include type definitions from. */
-    // "types": [],                           /* Type declaration files to be included in compilation. */
+    "strict": true,
```
make sure to make this false before opening the PR so that travis build ca pass

Ore4444 commented:
PR is clean now. Great work @hongaar 
I'm also asking myself what added values we gain from the interfaces

hongaar committed:
Merge remote-tracking branch 'upstream/develop' into feat/types

# Conflicts:
#	src/command.ts

hongaar committed:
chore: types

Some progress on typing Command. Removed autocomplete from types and moved to autocomplete module instead. Set bracketSpacing to default (having this at false really hurts my eyes while typing this code, I'm fine with setting it to false again later).

hongaar committed:
chore: types for Command

hongaar committed:
chore: rename types

hongaar committed:
chore: wip on typed EventEmitter

hongaar committed:
chore: wip on Vorpal.parse

hongaar committed:
chore: Vorpal.parse

hongaar committed:
chore: Vorpal.use

hongaar committed:
chore(eslint): --fix

hongaar committed:
chore: break down lodash import

hongaar committed:
refactor(ts): wip vorpal.command

hongaar committed:
refactor(ts): vorpal.command [wip]

hongaar committed:
refactor(ts): vorpal [wip]

hongaar committed:
refactor(ts): vorpal.prompt

hongaar committed:
refactor(ts): vorpal._prompt

hongaar committed:
refactor(ts): vorpal.exec

hongaar committed:
refactor(ts): vorpal._exec [wip]

hongaar committed:
refactor(ts): vorpal._exec [wip]

hongaar committed:
refactor(ts): command.description

hongaar committed:
chore(eslint): add lodash linter

AdrieanKhisbe commented on src/vorpal.ts:
```diff
@@ -2,11 +2,15 @@ import { EventEmitter } from 'events';
 
 import chalk from 'chalk';
-import _ from 'lodash';
+import _, {
```
Why do take both? :thinking:

hongaar replied:
Destructuring allows tree-shaking. But hope to factor out lodash entirely..!

AdrieanKhisbe replied:
ok, that's temporary

AdrieanKhisbe commented on .gitignore:
```diff
@@ -7,6 +7,7 @@ pids
 coverage
 .grunt
+.gitpod.yml
 dist/.local_storage*
```
By curiosité, whats that ? :detective:

hongaar replied:
Online version of VSCode (so I can do some stuff on my Chromebook while on the go). Switched to Visual Studio Online btw which turns out to be way better. Highly recommend it! 👍

AdrieanKhisbe replied:
I seee!
(I ended up there https://www.gitpod.io/)
Might give a try someday :)

AdrieanKhisbe commented on package.json:
```diff
@@ -66,9 +66,11 @@
     "eslint": "^6.6.0",
     "eslint-config-prettier": "^6.5.0",
     "eslint-plugin-prettier": "^3.1.1",
+    "eslint-plugin-you-dont-need-lodash-underscore": "^6.8.0",
```
Why not.
But I Hope it supports fp :)
(Will try it in m'y [branch](https://github.com/vorpaljs-reforged/vorpal/commits/feature/various-improvements) I havent yet made a PR out of.

AdrieanKhisbe commented on .eslintrc:
```diff
@@ -16,17 +16,21 @@
+        "@typescript-eslint/explicit-function-return-type": "off"
     }
-}
+}
```
Seems you dont like training New lines :wink:

hongaar replied:
I'm just lazy ;) Wonder why my editor doesn't take care of it though... must be a setting I can enable in vscode. 🤔

AdrieanKhisbe replied:
The laziest things would be to let the file as is ^^
Maybe thats your vscode configuration that make this? 🤔

AdrieanKhisbe commented on src/command-instance.ts:
```diff
@@ -13,15 +11,23 @@ interface CommandInstanceParams {
-  public command: any;
-  public session: any;
-  public parent: any;
+  public commandWrapper: CommandInstance;
```
:tada:

AdrieanKhisbe commented on src/command.ts:
```diff
@@ -1,89 +1,114 @@
+import Vorpal from './vorpal';
+
+export interface Args {
+  [key: string]: string | string[] | object | undefined;
```
Would bé interesting to have a (local?) type alias for thé values. (I mean for thé `s|s[]|o|u`

AdrieanKhisbe reviewed (commented):
Promising! :)
I just add a look of the start.

I'll try to make a real pass soon. (I started now from m'y phone, but it's too hell of a pain 🤭)

hongaar committed:
refactor(ts): vorpal._exec [wip]

hongaar committed:
refactor(ts): vorpal._exec

hongaar committed:
refactor(ts): vorpal._commandHelp

hongaar committed:
refactor(ts): vorpal._helpHeader

hongaar committed:
refactor(ts): vorpal

hongaar committed:
refactor(ts): session

hongaar committed:
refactor(ts): autocomplete

hongaar committed:
refactor(ts): command-instance / session

hongaar committed:
refactor(ts): history

hongaar committed:
refactor(ts): logger

hongaar committed:
refactor(ts): argh

hongaar committed:
refactor(ts): vorpal-commons

hongaar committed:
refactor(ts): option

hongaar committed:
refactor(ts): ui

hongaar committed:
refactor: ts

hongaar committed:
refactor: ts

hongaar commented:
Typing is done, tests are broken 😢 

This library needs very, very heavy refactoring. I still like the functionality and the general structure, but it's going to be a huge undertaking to make this code easier to work with...

hongaar committed:
chore: rm eslint-plugin-you-dont-need-lodash-underscore

hongaar committed:
chore: eslint fix

hongaar commented:
Typing is done, ~tests are~ everything is ~broken~ fixed again ~😢~ 🎉

This library needs very, very heavy refactoring. I still like the functionality and the general structure, but it's going to be a huge undertaking to make this code easier to work with.

Some untested behaviour might still be broken...

hongaar commented:
A lot of work went into this PR but unfortunately it seems we need a `vorpaljs-reforged-revived` initiative 😉 

Shameless plug: https://github.com/hongaar/bandersnatch
